### PR TITLE
fix(MemTable): make it cancel-safe and fix parallelism

### PR DESCRIPTION
**Pull request by DDtKey**
# Which issue does this PR close?

The issue is close to #5178, but for `MemTable`.

# Rationale for this change

1. It should be cancel-safe
2. There was an issue with join - `await` was sequential
3. There was possible to get a panic instead of Error
 

# What changes are included in this PR?

# Are these changes tested?

Existed tests are passing

# Are there any user-facing changes?

### CODE REVIEW

**Review by alamb — APPROVED** [review]
Thanks @DDtKey  ❤️ 

I left a suggestion that I think would make it even better but  I think this is an improvement (internal error vs panic 👍 ) already and could go in as is

File: datafusion/core/src/datasource/memory.rs [hunk 1]
```diff
@@ -76,20 +77,26 @@ impl MemTable {
...
             Vec::with_capacity(exec.output_partitioning().partition_count());
-        for task in tasks {
-            let result = task.await.expect("MemTable::load could not join task")?;
-            data.push(result);
+
+        for result in futures::future::join_all(tasks).await {
+            data.push(result.map_err(|_| {
+                DataFusionError::Internal(
+                    "MemTable::load could not join task".to_string(),
+                )
```
[the remark this answers is not in the document]
Comment (DDtKey) [reply datafusion/core/src/datasource/memory.rs:98]: `Context` requires `DataFusionError`, so it's anyway should be wrapped to something 🤔 

Could be `data.push(result.map_err(|e| DataFusionError::External(Box::new(e)))??)`, WDYT?

[the remark this answers is not in the document]
Comment (alamb) [reply datafusion/core/src/datasource/memory.rs:98]: That would definitely be better than 👍

[the remark this answers is not in the document]
Comment (DDtKey) [reply datafusion/core/src/datasource/memory.rs]: Done 🙂

**Review by alamb — APPROVED** [review]
Thanks again @DDtKey

**Pull request merged**